### PR TITLE
ext_authz: add dynamic_metadata_matchers and use them in the ext_authz HTTP implementation

### PR DESCRIPTION
**Pull request by esmet**
This allows for dynamic metadata when using an HTTP authorization service. Currently, it is only possible to set dynamic metadata using a gRPC authorization service.

Commit Message: ext_authz: add dynamic_metadata_matchers and use them in the ext_authz HTTP implementation
Additional Description:
Risk Level: low, new opt-in feature in an extension
Testing: unit tests
Docs Changes: proto definitions documented
Release Notes: ext_authz: added :ref:`dynamic_metadata_from_headers <envoy_v3_api_field_extensions.filters.http.ext_authz.v3.AuthorizationResponse.dynamic_metadata_from_headers>` to support emitting dynamic metadata from headers returned by an external authorization service via HTTP.

### CODE REVIEW

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc [hunk 1]
```diff
@@ -324,19 +342,24 @@ ResponsePtr RawHttpClientImpl::toResponse(Http::ResponseMessagePtr message) {
...
-        Response{CheckStatus::OK, Http::HeaderVector{}, Http::HeaderVector{}, Http::HeaderVector{},
-                 Http::HeaderVector{}, std::move(headers_to_remove), EMPTY_STRING, Http::Code::OK,
-                 ProtobufWkt::Struct{}}};
+    SuccessResponse ok{message->headers(),
+                       config_->upstreamHeaderMatchers(),
+                       config_->upstreamHeaderToAppendMatchers(),
+                       config_->clientHeaderOnSuccessMatchers(),
+                       config_->dynamicMetadataMatchers(),
+                       Response{CheckStatus::OK, Http::HeaderVector{}, Http::HeaderVector{},
+                                Http::HeaderVector{}, Http::HeaderVector{},
```
Comment (dio) [inline source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:351]: I'm not sure, while working with the other PR, I feel like we can have at least `CONSTRUCT_ON_FIRST_USE` for these `Http::HeaderVector{}`?

**Comment by dio** [comment]
@esmet Thanks for working on this. The implementation looks good. Do you think we can get the benefit from adding how this works in `docs/root/configuration/http/http_filters/ext_authz_filter.rst`? Probably in Dynamic Metadata section.

**Comment by wrowe** [comment]
/wait

on @esmet - ping for response to dio's comments above, otherwise LGTM

**Comment by esmet** [comment]
> @esmet Thanks for working on this. The implementation looks good. Do you think we can get the benefit from adding how this works in `docs/root/configuration/http/http_filters/ext_authz_filter.rst`? Probably in Dynamic Metadata section.

That makes sense. Let me take a look.

**Comment by esmet** [comment]
@wrowe I updated the docs to now accurately state that dynamic metadata is supported from an HTTP authorization server. Let me know what you think!

File: docs/root/configuration/http/http_filters/ext_authz_filter.rst [hunk 2]
```diff
@@ -179,16 +179,17 @@ Dynamic Metadata
...
-*only* when the gRPC authorization server returns a :ref:`CheckResponse
-<envoy_v3_api_msg_service.auth.v3.CheckResponse>` with a filled :ref:`dynamic_metadata
+When using a gRPC authorization server, dynamic metadata will be emitted only when the :ref:`CheckResponse
+<envoy_v3_api_msg_service.auth.v3.CheckResponse>` contains a filled :ref:`dynamic_metadata
 <envoy_v3_api_field_service.auth.v3.CheckResponse.dynamic_metadata>` field.
 
+When using an HTTP authorization server, dynamic metadata will be emitted only when there are response headers
+from the authorization server that match the configured
+:ref:`dynamic_metadata_from_headers <envoy_v3_api_field_extensions.filters.http.ext_authz.v3.AuthorizationResponse.dynamic_metadata_from_headers>`,
+if set. For every response header that matches, the filter will emit dynamic metadata whose key is the name of the matched header and whose value is the value of the matched header.
```
Comment (dio) [inline docs/root/configuration/http/http_filters/ext_authz_filter.rst:191]: Awesome, thanks!

**Comment by htuch** [comment]
@esmet can you update for merge conflict?

@wrowe @dio LMK when this is ready for senior maintainer review/merge, thanks!

File: docs/root/version_history/current.rst [hunk 3]
```diff
@@ -100,6 +100,7 @@ New Features
 * access_log: added :ref:`METADATA<envoy_v3_api_msg_extensions.formatter.metadata.v3.Metadata>` token to handle all types of metadata (DYNAMIC, CLUSTER, ROUTE).
 * bootstrap: added :ref:`inline_headers <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.inline_headers>` in the bootstrap to make custom inline headers bootstrap configurable.
 * contrib: added new :ref:`contrib images <install_contrib>` which contain contrib extensions.
+* ext_authz: added :ref:`dynamic_metadata_from_headers <envoy_v3_api_field_extensions.filters.http.ext_authz.v3.AuthorizationResponse.dynamic_metadata_from_headers>` to support emitting dynamic metadata from headers returned by an external authorization service via HTTP.
```
Comment (dio) [inline docs/root/version_history/current.rst:103]: Sorry, do you mean: HTTP-based (I'm also not sure for phrasing this right when trying to say the HTTP implementation (vs. gRPC-based)) external authorization service?

**Comment by dio** [comment]
@htuch, once @esmet resolves the merge conflict, I think this is good to go (I have a question https://github.com/envoyproxy/envoy/pull/17895#discussion_r704059886 on that `Http::HeaderVector{}` initialization, but that's minor).

File: source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc (hunk 1 above)
Comment (esmet) [reply source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:351]: Sounds reasonable. Let's address this in a new PR?

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by htuch** [comment]
@esmet one more conflict to resolve, thanks.
/wait

**Comment by esmet** [comment]
@htuch updated! :)

**Review by htuch — APPROVED** [review]

File: docs/root/version_history/current.rst (hunk 3 above)
Comment (esmet) [reply docs/root/version_history/current.rst:108]: I missed this comment - my bad. Yes, this means an HTTP-based authorization service.

**Pull request merged**